### PR TITLE
ocamlPackages.ctypes: 0.17.1 -> 0.18.0

### PR DESCRIPTION
**Pull request by sternenseemann**
<https://github.com/ocamllabs/ocaml-ctypes/blob/0.18.0/CHANGES.md#ctypes-0180>

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 115290` at 419b4c4e run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/1d53733b8f618f660a84ad865b91a814">libbap</a></li></li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/bf1596ee8f8b06ef5016628e2aca0481">liquidsoap</a></li></li>
  </ul>
</details>

**Comment by vbgl** [comment]
@GrahamcOfBorg build ocamlPackages.torch

File: pkgs/development/libraries/libbap/default.nix [hunk 1]
```diff
@@ -12,6 +12,10 @@ stdenv.mkDerivation {
     sha256 = "0m4spva3z6fgbwlg4zq53l5p227dic893q2qq65pvzxyf7k7nmil";
   };
 
+  postPatch = ''
+    sed -i 's/-linkpkg/-thread -linkpkg/g' Makefile.in
```
Comment (vbgl) [inline pkgs/development/libraries/libbap/default.nix:16]: What about using `substituteInPlace`?

File: pkgs/development/ocaml-modules/janestreet/0.14.nix [hunk 2]
```diff
@@ -147,6 +147,10 @@ rec {
     meta.description = "Async wrappers for SSL";
     buildInputs = [ dune-configurator ];
     propagatedBuildInputs = [ async ctypes openssl ];
+    # in ctypes.foreign 0.18.0 threaded and unthreaded have been merged
+    postPatch = ''
+      sed -i 's/ctypes.foreign.threaded/ctypes.foreign/g' bindings/dune
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/janestreet/0.14.nix:152]: Same question about `substituteInPlace`.

**Review by vbgl — CHANGES_REQUESTED** [review]
This seems to break a few things (e.g., `ocaml-ng.ocamlPackages_4_07.sodium`).

**Comment by sternenseemann** [comment]
> This seems to break a few things (e.g., ocaml-ng.ocamlPackages_4_07.sodium).

Good, catch, I'll investigate. Previously I only had `torch` on my list, but for OCaml 4.10 it compiles successfully for me.

**Comment by sternenseemann** [comment]
@ofborg build ocamlPackages.torch

**Comment by sternenseemann** [comment]
@ofborg build ocaml-ng.ocamlPackages_4_07.sodium

**Comment by sternenseemann** [comment]
@ofborg build ocamlPackages.torch

**Comment by sternenseemann** [comment]
`ocamlPackages.torch` is broken on master as well, seems to be a fairly recent breakage as well. Currently bisecting to find out what happened, but this is gonna take a while if I'm gonna hit more `pytorch` rebuilds…

**Comment by vbgl** [comment]
The `torch` issue is unrelated to this pr. Hint: #115173.

**Comment by sternenseemann** [comment]
> The `torch` issue is unrelated to this pr. Hint: #115173.

Seems like there is already a fix and hopefully a new release soon-ish: <https://github.com/LaurentMazare/ocaml-torch/commit/21b87834e797958f0aea1bbf0a8ddccdf995fbe3>

**Comment by vbgl** [comment]
This breaks `ocaml-ng.ocamlPackages_4_04.async_ssl` (removing this legacy package might be an acceptable fix).

**Comment by sternenseemann** [comment]
> This breaks `ocaml-ng.ocamlPackages_4_04.async_ssl` (removing this legacy package might be an acceptable fix).

Seems like it breaks every package built using dune 1 which depends on `ctypes` due to `bigarray-compat`.

**Pull request merged**